modvendor/modvendor_test.go

```diff
@@ -15,18 +15,823 @@
+//
+// Example:
+//   git::file://{{.}}/module-test?ref=main#main.tf
+type vendorPathSpec string
+
+type testcase struct {
+	name         string
+	source       string
+	layout       []string
+	vendordir    string
+	configs      []hclconfig
+	wantVendored []string
+	wantIgnored  []wantIgnoredVendor
+	wantError    error
+
+	// wantFiles describes file contents that must exist in the vendor directory.
+	// The fmt.Stringer can be a template that gets compiled with text/template
+	// package and having access to the relative paths of the vendored modules.
+	// For example, if `wantVendored` contains [<modsrc1>, <modsrc2>, ...], then
+	// the template will have access to the list [<relative path of modsrc1>,
+	// <relative path of modsrc2>, ...].
+	// Then the template can reference the third element of the list:
+	//   source = {{index 2}}
+	// If this looks like a hack to you, it's because it is.
+	wantFiles map[vendorPathSpec]fmt.Stringer
+}
+
+type hclconfig struct {
+	repo string
+	path string
+	data fmt.Stringer
+}
+
+type wantIgnoredVendor struct {
+	RawSource     string
+	ReasonPattern string
+}
+
+type wantReport struct {
+	Vendored []string
+	Ignored  []wantIgnoredVendor
+	Error    error
+}
+
+func TestModVendorAllRecursive(t *testing.T) {
+	tcases := []testcase{
+		{
+			name: "module with no remote deps",
+			layout: []string{
+				"g:module-test",
+			},
+			source: "git::file://{{.}}/module-test?ref=main",
+			configs: []hclconfig{
+				{
+					repo: "module-test",
+					path: "module-test/main.tf",
+					data: Module(
+						Labels("test"),
+						Str("source", "./modules/non-existent"),
+					),
+				},
+			},
+			wantVendored: []string{
+				"git::file://{{.}}/module-test?ref=main",
+			},
+		},
+		{
+			name: "module with HCL errors",
+			layout: []string{
+				"g:module-test",
+			},
+			source: "git::file://{{.}}/module-test?ref=main",
+			configs: []hclconfig{
+				{
+					repo: "module-test",
+					path: "module-test/main.tf",
+					data: bytes.NewBufferString("this is not a valid HCL file"),
+				},
+			},
+			wantVendored: []string{
+				"git::file://{{.}}/module-test?ref=main",
+			},
+			wantError: errors.E(hcl.ErrHCLSyntax),
+		},
+		{
+			name: "module with ignored remote deps",
+			layout: []string{
+				"g:module-test",
+			},
+			source: "git::file://{{.}}/module-test?ref=main",
+			configs: []hclconfig{
+				{
+					repo: "module-test",
+					path: "module-test/main.tf",
+					data: Module(
+						Labels("test"),
+						Str("source", "https://example.com/my-module"),
+					),
+				},
+			},
+			wantVendored: []string{
+				"git::file://{{.}}/module-test?ref=main",
+			},
+			wantIgnored: []wantIgnoredVendor{
+				{
+					RawSource:     "https://example.com/my-module",
+					ReasonPattern: "unsupported module source",
+				},
+			},
+		},
+		{
+			name:   "module not found",
+			source: "git::file://{{.}}/module-that-does-not-exists?ref=main",
+			wantIgnored: []wantIgnoredVendor{
```

Comment:
hmm if the reference is git but doesn't exist it is also just ignored ?

Reply:
the problem is that we have to try to clone() in order to know the issues (not exists, ref doesn't exists, etc), so yeah now it returns in the `ignored` list but we can improve later.